nomos-blend/message/src/encap.rs

```diff
@@ -0,0 +1,481 @@
+use std::iter::repeat_n;
+
+use nomos_core::wire;
+use serde::{Deserialize, Serialize};
+
+use crate::{
+    crypto::{
+        pseudo_random_bytes, Ed25519PrivateKey, Ed25519PublicKey, ProofOfQuota, ProofOfSelection,
+        SharedKey, Signature, X25519PrivateKey, KEY_SIZE, PROOF_OF_QUOTA_SIZE,
+        PROOF_OF_SELECTION_SIZE, SIGNATURE_SIZE,
+    },
+    error::Error,
+    message::{BlendingHeader, Header, Payload, PayloadHeader, PayloadType, PublicHeader},
+};
+
+// TODO: Consider having these as parameters or generics.
+const MAX_ENCAPSULATIONS: usize = 3;
+const MAX_PAYLOAD_BODY_SIZE: usize = 34 * 1024;
+
+/// An encapsulated message that is sent to the blend network.
+//
+// TODO: Implement [`BlendMessage`] trait for this type.
+#[derive(Clone, Serialize, Deserialize)]
+pub struct EncapsulatedMessage {
+    /// A header that is not encapsulated.
+    header: Header,
+    /// A public header that is not encapsulated.
+    public_header: PublicHeader,
+    /// An encapsulated private header.
+    private_header: [Vec<u8>; MAX_ENCAPSULATIONS],
+    /// A encapsulated payload.
+    payload: Vec<u8>,
+}
+
+impl EncapsulatedMessage {
+    /// Encapsulates the given payload with the provided encapsulation inputs.
+    /// Encapsulation inputs more than [`MAX_ENCAPSULATIONS`] are not allowed.
+    pub fn new(
+        inputs: &[EncapsulationInput],
+        payload_type: PayloadType,
+        payload_body: &[u8],
+    ) -> Result<Self, Error> {
+        if inputs.len() > MAX_ENCAPSULATIONS {
+            return Err(Error::MaxEncapsulationsExceeded);
+        }
+        if inputs.is_empty() {
+            return Err(Error::EmptyEncapsulationInputs);
+        }
+        if payload_body.len() > MAX_PAYLOAD_BODY_SIZE {
+            return Err(Error::PayloadTooLarge);
+        }
+
+        let msg = Self::initialize(inputs, payload_type, payload_body)?;
+
+        // Encapsulate message using each shared key.
+        let (mut msg, _) = inputs.iter().fold((msg, None), |(msg, prev_input), input| {
+            let msg = Self::encapsulate(msg, input, prev_input);
+            (msg, Some(input))
+        });
+
+        // Construct the public header.
+        let EncapsulationInput {
+            signing_key,
+            proof_of_quota,
+            ..
+        } = inputs.last().expect("inputs should not be empty");
+        let sig = signing_key.sign(&msg.signing_body());
+        msg.public_header = PublicHeader {
+            signing_pubkey: signing_key.public_key(),
+            proof_of_quota: proof_of_quota.clone(),
+            signature: sig,
+        };
+
+        Ok(msg)
+    }
+
+    /// An initialization step for preparing encapsulations.
+    fn initialize(
```

Comment:
I think this would be a perfect fit for the builder/typestate pattern. Each step of the encapsulation has a different type, on which you can call a specific set of functions to get something else in return. So here we can have an uninitialised and an initialised message type, which contain different data. These are nits, especially given this is early code and the time constraints, but wanted to raise my point.